[C++ frontend] error: no member named 'is_sorted' in namespace 'std'

````CPP
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cassert>

class Car {
public:
    std::string brand;
    int speed;

    Car(std::string brand, int speed) : brand(brand), speed(speed) {}
};

bool compare_speed(const Car& c1, const Car& c2) {
    return c1.speed < c2.speed;
}

int main() {
    std::vector<Car> cars = {
        Car("Toyota", 120),
        Car("Honda", 130),
        Car("Ford", 110)
    };

    std::vector<Car> sorted_cars = cars;
    std::sort(sorted_cars.begin(), sorted_cars.end(), compare_speed);

    assert(std::is_sorted(sorted_cars.begin(), sorted_cars.end(), compare_speed) && "List is not sorted correctly!");

    std::cout << "All assertions passed successfully!" << std::endl;
    return 0;
}
````

````
$ esbmc lambda.cpp
ESBMC version 7.8.1 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmclibc
Parsing lambda.cpp
lambda.cpp:29:17: error: no member named 'is_sorted' in namespace 'std'
    assert(std::is_sorted(sorted_cars.begin(), sorted_cars.end(), compare_speed) && "List is not sorted correctly!");
           ~~~~~^
/usr/include/assert.h:93:27: note: expanded from macro 'assert'
     (static_cast <bool> (expr)                                         \
                          ^~~~
In file included from lambda.cpp:2:
/tmp/esbmc-cpp-headers-7e76-c904-cc50/vector:202:15: error: no matching constructor for initialization of 'Car[]'
    buf = new T[init.size()];
              ^
lambda.cpp:20:29: note: in instantiation of member function 'std::vector<Car>::vector' requested here
    std::vector<Car> cars = {
                            ^
lambda.cpp:7:7: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 0 were provided
class Car {
      ^
lambda.cpp:7:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 0 were provided
lambda.cpp:12:5: note: candidate constructor not viable: requires 2 arguments, but 0 were provided
    Car(std::string brand, int speed) : brand(brand), speed(speed) {}
    ^
In file included from lambda.cpp:2:
/tmp/esbmc-cpp-headers-7e76-c904-cc50/vector:192:15: error: no matching constructor for initialization of 'Car[]'
    buf = new T[x._size];
              ^
lambda.cpp:26:36: note: in instantiation of member function 'std::vector<Car>::vector' requested here
    std::vector<Car> sorted_cars = cars;
                                   ^
lambda.cpp:7:7: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 0 were provided
class Car {
      ^
lambda.cpp:7:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 0 were provided
lambda.cpp:12:5: note: candidate constructor not viable: requires 2 arguments, but 0 were provided
    Car(std::string brand, int speed) : brand(brand), speed(speed) {}
    ^
ERROR: PARSING ERROR
````

--- a/regression/esbmc-cpp/algorithm/algorithm101_partial_sort/test.desc
+++ b/regression/esbmc-cpp/algorithm/algorithm101_partial_sort/test.desc
@@ -1,4 +1,4 @@
 CORE
 main.cpp
---unwind 10 --no-unwinding-assertions  --z3 --timeout 900
+--unwind 10 --no-unwinding-assertions --timeout 900
 ^VERIFICATION FAILED$

--- a/regression/esbmc-cpp/algorithm/algorithm102_partial_sort_copy/test.desc
+++ b/regression/esbmc-cpp/algorithm/algorithm102_partial_sort_copy/test.desc
@@ -1,4 +1,4 @@
 CORE
 main.cpp
---unwind 10 --no-unwinding-assertions --timeout 900
+--unwind 7 --no-unwinding-assertions --timeout 900
 ^VERIFICATION FAILED$

--- a/regression/esbmc-cpp/vector/vector63/test.desc
+++ b/regression/esbmc-cpp/vector/vector63/test.desc
@@ -1,4 +1,4 @@
 CORE
 main.cpp
-
+--incremental-bmc
 ^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp/vector/vector64/main.cpp
+++ b/regression/esbmc-cpp/vector/vector64/main.cpp
@@ -1,0 +1,33 @@
+#include <iostream>
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <cassert>
+
+class Car {
+public:
+    std::string brand;
+    int speed;
+
+    Car(std::string brand, int speed) : brand(brand), speed(speed) {}
+};
+
+bool compare_speed(const Car& c1, const Car& c2) {
+    return c1.speed < c2.speed;
+}
+
+int main() {
+    std::vector<Car> cars = {
+        Car("Toyota", 120),
+        Car("Honda", 130),
+        Car("Ford", 110)
+    };
+
+    std::vector<Car> sorted_cars = cars;
+    std::sort(sorted_cars.begin(), sorted_cars.end(), compare_speed);
+
+    //assert(std::is_sorted(sorted_cars.begin(), sorted_cars.end(), compare_speed) && "List is not sorted correctly!");
+
+    std::cout << "All assertions passed successfully!" << std::endl;
+    return 0;
+}

--- a/regression/esbmc-cpp/vector/vector64/test.desc
+++ b/regression/esbmc-cpp/vector/vector64/test.desc
@@ -1,4 +1,4 @@
 CORE
 main.cpp
-
+--unwind 5 --no-unwinding-assertions
 ^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp/vector/vector64/test.desc
+++ b/regression/esbmc-cpp/vector/vector64/test.desc
@@ -1,4 +1,4 @@
-CORE
+KNOWNBUG
 main.cpp
 --unwind 5 --no-unwinding-assertions
 ^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp/vector/vector64/test.desc
+++ b/regression/esbmc-cpp/vector/vector64/test.desc
@@ -1,4 +1,4 @@
-KNOWNBUG
+CORE
 main.cpp
 --unwind 5 --no-unwinding-assertions
 ^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp/vector/vector65/main.cpp
+++ b/regression/esbmc-cpp/vector/vector65/main.cpp
@@ -18,16 +18,17 @@ bool compare_speed(const Car& c1, const Car& c2) {
 
 int main() {
     std::vector<Car> cars = {
-        Car("Toyota", 120),
-        Car("Honda", 130),
-        Car("Ford", 110)
+        Car("T", 200),
+        Car("H", 130),
+        Car("F", 110)
     };
 
     std::vector<Car> sorted_cars = cars;
-    std::sort(sorted_cars.begin(), sorted_cars.end(), compare_speed);
+    // do not sort to make the assertion to fail!
+    //std::sort(sorted_cars.begin(), sorted_cars.end(), compare_speed);
 
     assert(std::is_sorted(sorted_cars.begin(), sorted_cars.end(), compare_speed) && "List is not sorted correctly!");
 
-    std::cout << "Assertion passed successfully!" << std::endl;
+    std::cout << "Assertion should fail!" << std::endl;
     return 0;
 }

--- a/regression/esbmc-cpp/vector/vector65/test.desc
+++ b/regression/esbmc-cpp/vector/vector65/test.desc
@@ -1,4 +1,4 @@
-KNOWNBUG
+CORE
 main.cpp
 --unwind 5 --no-unwinding-assertions --no-pointer-check
 ^VERIFICATION FAILED$

--- a/regression/esbmc-cpp/vector/vector65/test.desc
+++ b/regression/esbmc-cpp/vector/vector65/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.cpp
+--unwind 5 --no-unwinding-assertions --no-pointer-check
+^VERIFICATION FAILED$

--- a/regression/esbmc-cpp/vector/vector65/test.desc
+++ b/regression/esbmc-cpp/vector/vector65/test.desc
@@ -1,4 +1,4 @@
-CORE
+KNOWNBUG
 main.cpp
 --unwind 5 --no-unwinding-assertions --no-pointer-check
 ^VERIFICATION FAILED$

--- a/regression/esbmc-cpp/vector/vector67/main.cpp
+++ b/regression/esbmc-cpp/vector/vector67/main.cpp
@@ -1,0 +1,19 @@
+#include <vector>
+
+class Car {
+public:
+    int brand;
+    int speed;
+
+    Car(int brand, int speed) : brand(brand), speed(speed) {}
+};
+
+int main() {
+    std::vector<Car> cars = {
+        Car(1, 120),
+        Car(2, 130),
+        Car(3, 110)
+    };
+
+    return 0;
+}

--- a/regression/esbmc-cpp/vector/vector67/test.desc
+++ b/regression/esbmc-cpp/vector/vector67/test.desc
@@ -1,0 +1,4 @@
+KNOWNBUG
+main.cpp
+--incremental-bmc --memory-leak-check
+^VERIFICATION SUCCESSFUL$

--- a/src/cpp/library/algorithm
+++ b/src/cpp/library/algorithm
@@ -9,6 +9,32 @@ typedef bool(function)(int arg1, int arg2);
 
 namespace std
 {
+template<class ForwardIt, class Compare>
+bool is_sorted(ForwardIt first, ForwardIt last, Compare comp) {
+    if (first != last) {
+        ForwardIt next = first;
+        while (++next != last) {
+            if (comp(*next, *first))
+                return false;
+            first = next;
+        }
+    }
+    return true;
+}
+
+template<class ForwardIt>
+bool is_sorted(ForwardIt first, ForwardIt last) {
+    if (first != last) {
+        ForwardIt next = first;
+        while (++next != last) {
+            if (*next < *first)
+                return false;
+            first = next;
+        }
+    }
+    return true;
+}
+
 template <class InputIterator, class Function>
 Function for_each(InputIterator first, InputIterator last, Function f)
 {

--- a/src/cpp/library/vector
+++ b/src/cpp/library/vector
@@ -157,7 +157,8 @@ public:
     for (; tmp != t2; tmp++)
       count++;
 
-    buf = new T[count];
+    buf = static_cast<T*>(malloc(sizeof(T) * count));
+    __ESBMC_assume(buf);
     int n = 0;
     for (; t1 != t2; t1++)
     {
@@ -186,7 +187,8 @@ public:
     for (InputIt tmp = t; tmp != addr; tmp++)
       count++;
 
-    buf = new T[count];
+    buf = static_cast<T*>(malloc(sizeof(T) * count));
+    __ESBMC_assume(buf);
     for (int i = 0; i < count; i++)
       buf[i] = *t++;
 
@@ -206,7 +208,8 @@ public:
 
   vector(const vector &x) : _size(0), _capacity(1)
   {
-    buf = new T[x._size];
+    buf = static_cast<T*>(malloc(sizeof(T) * x._size));
+    __ESBMC_assume(buf);
     _size = x._size;
     for (int i = 0; i < _size; i++)
       buf[i] = x.buf[i];
@@ -227,7 +230,8 @@ public:
 
   vector &operator=(const vector &x)
   {
-    buf = new T[x._size];
+    buf = static_cast<T*>(malloc(sizeof(T) * x._size));
+    __ESBMC_assume(buf);
     _size = x._size;
     for (int i = 0; i < _size; i++)
       buf[i] = x.buf[i];
@@ -252,7 +256,8 @@ public:
     for (; tmp != t2; tmp++)
       count++;
 
-    buf = new T[count];
+    buf = static_cast<T*>(malloc(sizeof(T) * count));
+    __ESBMC_assume(buf);
     int n = 0;
     for (; t1 != t2; t1++)
     {

--- a/src/cpp/library/vector
+++ b/src/cpp/library/vector
@@ -135,6 +135,8 @@ public:
     _size = 0;
     _capacity = 10;
   }
+  // TODO: Support destructors. 
+  // Check discussion at https://github.com/esbmc/esbmc/pull/2515
 #if 0
   // Destructor
   ~vector()

--- a/src/cpp/library/vector
+++ b/src/cpp/library/vector
@@ -155,8 +155,7 @@ public:
     for (; tmp != t2; tmp++)
       count++;
 
-    buf = static_cast<T*>(malloc(sizeof(T) * count));
-    __ESBMC_assume(buf);
+    buf = new T[count];
     int n = 0;
     for (; t1 != t2; t1++)
     {
@@ -206,8 +205,7 @@ public:
 
   vector(const vector &x) : _size(0), _capacity(1)
   {
-    buf = static_cast<T*>(malloc(sizeof(T) * x._size));
-    __ESBMC_assume(buf);
+    buf = new T[x._size];
     _size = x._size;
     for (int i = 0; i < _size; i++)
       buf[i] = x.buf[i];
@@ -228,8 +226,7 @@ public:
 
   vector &operator=(const vector &x)
   {
-    buf = static_cast<T*>(malloc(sizeof(T) * x._size));
-    __ESBMC_assume(buf);
+    buf = new T[x._size];
     _size = x._size;
     for (int i = 0; i < _size; i++)
       buf[i] = x.buf[i];

--- a/src/cpp/library/vector
+++ b/src/cpp/library/vector
@@ -186,8 +186,7 @@ public:
     for (InputIt tmp = t; tmp != addr; tmp++)
       count++;
 
-    buf = static_cast<T*>(malloc(sizeof(T) * count));
-    __ESBMC_assume(buf);
+    buf = new T[count];
     for (int i = 0; i < count; i++)
       buf[i] = *t++;
 
@@ -252,9 +251,8 @@ public:
     iterator tmp = t1;
     for (; tmp != t2; tmp++)
       count++;
-    
-    buf = static_cast<T*>(malloc(sizeof(T) * count));
-    __ESBMC_assume(buf);
+
+    buf = new T[count];
     int n = 0;
     for (; t1 != t2; t1++)
     {

--- a/src/cpp/library/vector
+++ b/src/cpp/library/vector
@@ -130,11 +130,24 @@ public:
   // construct:
   explicit vector() : _size(0), _capacity(1)
   {
-    buf = new T[10];
+    buf = static_cast<T*>(malloc(sizeof(T) * 10));
+    __ESBMC_assume(buf);
     _size = 0;
     _capacity = 10;
   }
-
+#if 0
+  // Destructor
+  ~vector()
+  {
+    if (buf != nullptr)
+    {
+      free(buf);
+      buf = nullptr;
+      _size = 0;
+      _capacity = 0;
+    }
+  }
+#endif
   explicit vector(iterator t1, iterator t2) : _size(0)
   {
     int count = 0;
@@ -142,7 +155,8 @@ public:
     for (; tmp != t2; tmp++)
       count++;
 
-    buf = new T[count];
+    buf = static_cast<T*>(malloc(sizeof(T) * count));
+    __ESBMC_assume(buf);
     int n = 0;
     for (; t1 != t2; t1++)
     {
@@ -154,7 +168,8 @@ public:
 
   explicit vector(size_type n) : _size(0), _capacity(1)
   {
-    buf = new T[n];
+    buf = static_cast<T*>(malloc(sizeof(T) * n));
+    __ESBMC_assume(buf);
     _size = 0;
     for (int i = 0; i < n; i++)
     {
@@ -170,7 +185,8 @@ public:
     for (InputIt tmp = t; tmp != addr; tmp++)
       count++;
 
-    buf = new T[count];
+    buf = static_cast<T*>(malloc(sizeof(T) * count));
+    __ESBMC_assume(buf);
     for (int i = 0; i < count; i++)
       buf[i] = *t++;
 
@@ -180,7 +196,8 @@ public:
 
   explicit vector(size_type n, const T &x)
   {
-    buf = new T[n];
+    buf = static_cast<T*>(malloc(sizeof(T) * n));
+    __ESBMC_assume(buf);
     _size = n;
     for (int i = 0; i < n; i++)
       buf[i] = x;
@@ -203,15 +220,16 @@ public:
     buf = static_cast<T*>(malloc(sizeof(T) * init.size()));
     __ESBMC_assume(buf);
     _size = init.size();
-    auto it = init.begin();
-    for (int i = 0; i < _size; i++, ++it)
-      new(&buf[i]) T(*it);
+    for (int i = 0; i < _size; i++)
+      buf[i] = *(init.begin() + i);
+    _capacity = _size;
   }
 #endif
 
   vector &operator=(const vector &x)
   {
-    buf = new T[x._size];
+    buf = static_cast<T*>(malloc(sizeof(T) * x._size));
+    __ESBMC_assume(buf);
     _size = x._size;
     for (int i = 0; i < _size; i++)
       buf[i] = x.buf[i];
@@ -235,8 +253,9 @@ public:
     iterator tmp = t1;
     for (; tmp != t2; tmp++)
       count++;
-
-    buf = new T[count];
+    
+    buf = static_cast<T*>(malloc(sizeof(T) * count));
+    __ESBMC_assume(buf);
     int n = 0;
     for (; t1 != t2; t1++)
     {

--- a/src/cpp/library/vector
+++ b/src/cpp/library/vector
@@ -189,7 +189,8 @@ public:
 
   vector(const vector &x) : _size(0), _capacity(1)
   {
-    buf = new T[x._size];
+    buf = static_cast<T*>(malloc(sizeof(T) * x._size));
+    __ESBMC_assume(buf);
     _size = x._size;
     for (int i = 0; i < _size; i++)
       buf[i] = x.buf[i];
@@ -199,11 +200,12 @@ public:
 #if __cplusplus >= 201103L
   vector(std::initializer_list<T> init) :_size(0), _capacity(1)
   {
-    buf = new T[init.size()];
+    buf = static_cast<T*>(malloc(sizeof(T) * init.size()));
+    __ESBMC_assume(buf);
     _size = init.size();
-    for (int i = 0; i < _size; i++)
-      buf[i] = *(init.begin() + i);
-    _capacity = _size;
+    auto it = init.begin();
+    for (int i = 0; i < _size; i++, ++it)
+      new(&buf[i]) T(*it);
   }
 #endif
 

--- a/src/goto-symex/builtin_functions.cpp
+++ b/src/goto-symex/builtin_functions.cpp
@@ -267,7 +267,7 @@ expr2tc goto_symext::symex_mem(
   symex_assign(code_assign2tc(lhs, rhs), true);
 
   expr2tc ptr_obj = pointer_object2tc(pointer_type2(), ptr_rhs);
-  track_new_pointer(ptr_obj, new_type);
+  track_new_pointer(ptr_obj, new_type, size);
 
   dynamic_memory.emplace_back(
     rhs_copy, alloc_guard, !is_malloc, symbol.name.as_string());

